enhancements/sig-architecture/224-event-based-manifestwork/README.md

```diff
@@ -0,0 +1,788 @@
+    // <cluster-id> represents the identifier of the target cluster to which this cloud event is sent.
+    // <resource-type> refers to the resource included in the cloud event payload.
+    "source": "/sources/<source-id>/clusters/<cluster-id>/<resource-type>",
+    // Use "<reverse-domain-name-of-resource-organization>.<resource-version>.<resource-name>.<subsource>.<action>" as the format for this attribute, which contains a value describing the type of event related to the originating occurrence.
+    // For example, "io.open-cluster-management.v1.manifestwork.statusresync.requested" can be used as a sample format.
+    "type": "<reverse-domain-name-of-resource-organization>.<resource-version>.<resource-name>.<subsource>.<action>",
+    // Timestamp of when the occurrence happened. Optional, if present, MUST adhere to the format specified in RFC 3339.
+    "time": "<timestamp-of-the-event-occurrence>",
+    // Content type of data value, If present, MUST adhere to the format specified in RFC 2046.
+    // By default "application/json" will be used if it is not set.
+    "datacontenttype": "<data-content-type>"
+}
+// Data
+{
+    "statusHashes": [
+        {
+        "resourceID": "748d04ad-887a-43aa-b123-da1373227011",
+        "statusHash": "12345abc"
+        },
+        {
+        "resourceID":  "858d03ad-997b-52bc-b456-ab1473228022",
+        "statusHash": "abc12345"
+        }
+    ]
+}
+```
+
+### Failure Handling
+
+~~- Retry Mechanism(TBD):~~
+
+~~In the event of a failure during event delivery or processing, a retry mechanism should be implemented to ensure reliable event propagation. When an event fails to reach its destination or encounters an error during processing, it should be retried for a certain number of times before giving up.~~
+
+~~The retry mechanism should employ exponential backoff, gradually increasing the delay between retries to prevent overwhelming the system with failed events. Each retry attempt should be logged, including relevant information such as the event payload, timestamp, and error details.~~
+
+- Resync Mechanism:
+
+To ensure the integrity of the event-based manifestworks system during source or work agent restart or reconnection, the following optimized resync mechanism is employed:
+
+1. On start, the source subscribes to the spec resync event topic /sources/resync/+/manifests from the broker.
+2. On start, the work agent subscribes to the status resync event topic /sources/+/resync/clusters/manifestsstatus from the broker.
+3. Upon work agent restart or reconnect, the work agent generates a spec resync event with resource versions and sends it to the broker.
+4. Upon source restart or reconnect, the source calculates a status hash of its owned resources and sends a status resync event with the resource hash to the broker.
```

Comment:
AFAIK, MQTT broker has the framework to handle 1,2,3 & 4. We should not need any code to do it... beyond changing configurations in MQTT connection.

Reply:
`resync` need to handle by our code. MQTT can retain message, but it has limitation. that is why we need to handle `resync` by ourselves.

Reply:
yes, we may leverage the MQTT session mechanism to relieve this problem, but it still cannot avoid this totally, e.g. when the agent connect the broker after its session is expiration, the agent will lost the message, and even if we set the session will never expire (this will occupy the more broker resources), if the broker is crash, the source/agent  is still possible to lose messages, so we still need the resync

Reply:
`resync` is part of the communication protocol, while the transport configuration, is it part of the protocol? Should we let the user to enable message retain?

Reply:
@skeeey you hit the nail on the head. We should leverage MQTT session mechanism, set the session will never expire and assume that broker does not lose messages. And then see what happens before adding more code IMO. We should assume that broker should be resilient - as they are today. WDYT?
In Global Hub we added a powerful concept of sending delta message events and periodic complete events. That is not relevant here - looking at the current events being exchanged.

Reply:
@morvencao how does source `calculates a status hash of its owned resources and sends a status resync event with the resource hash to the broker`  if its memory is empty on restart (assuming this is all stored in memory)

Reply:
If the source is restful service with DB, then the status can be from DB. If the source is hub controller, then emty hash slice will be sent to agent.

Reply:
Are you saying resync operation does not do anything useful for hub controllers?

Reply:
no, for empty status hash, agent will send back all the status back to source.

Reply:
I think resync can be optional if the broker implementation has other way to handle the drift. But we should still need resync as the last resort.

Reply:
@bjoydeep, yeah, I think we just provide a resync mechanism here (we do not make any assumptions about the broker), and we have a client to provide the pub/sub/resync functions, for deveployers/users, if they have a reliable
 broker, they can choose not to do resync between sources and agents

Reply:
also we can not make an assumption on who stores the source of truth, it could be transport layer or it could be the consumer, resync ensures the agent can get the source of truth ultimately.

Reply:
Agreed.